src/main.cpp

```diff
@@ -2145,18 +2238,42 @@ void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo &txund
             CCoinsModifier coins = inputs.ModifyCoins(txin.prevout.hash);
             unsigned nPos = txin.prevout.n;
 
-            if (nPos >= coins->vout.size() || coins->vout[nPos].IsNull())
-                assert(false);
+            assert(nPos < coins->vout.size() && !coins->vout[nPos].IsNull());
+
             // mark an outpoint spent, and construct undo information
             txundo.vprevout.push_back(CTxInUndo(coins->vout[nPos]));
             coins->Spend(nPos);
-            if (coins->vout.size() == 0) {
+
+            // If this was the last unspent output, save information
+            // about the transaction along with the undo for this output
+            // so that we can restore it when unwinding.
+            if (!coins->HasUnspent()) {
                 CTxInUndo& undo = txundo.vprevout.back();
                 undo.nHeight = coins->nHeight;
                 undo.fCoinBase = coins->fCoinBase;
                 undo.nVersion = coins->nVersion;
             }
         }
+
+        txundo.vtzeprevout.reserve(tx.vtzein.size());
+        for (const CTzeIn &tzein : tx.vtzein) {
+            CCoinsModifier coins = inputs.ModifyCoins(tzein.prevout.hash);
+            unsigned nPos = tzein.prevout.n;
+
+            assert(nPos < coins->vtzeout.size() && coins->vtzeout[nPos].second != SPENT);
+
+            txundo.vtzeprevout.push_back(CTzeInUndo(coins->vtzeout[nPos].first, coins->nHeight, coins->nVersion));
+            coins->SpendTzeOut(nPos);
+
+            // If this was the last unspent output, save information
+            // about the transaction along with the undo for this output
+            // so that we can restore it when unwinding.
+            if (!coins->HasUnspent()) {
+                CTzeInUndo& undo = txundo.vtzeprevout.back();
+                undo.nHeight = coins->nHeight;
+                undo.nVersion = coins->nVersion;
+            }
```

Comment:
This logic is setting the `nHeight` and `nVersion` fields of the `CTzeInUndo` to values regardless of whether `!coins->HasUnspent()` is true later, violating the documentation

```
    // The following fields are only in the case that the associated output is the last
    // unspent output (transparent or TZE) of a transaction. Since blocks are unwound
    // from tip to base, at the time that such an undo is applied this information can
    // be used to reconstruct a CCoins value that caches unspent outputs for the transaction.
    unsigned int nHeight; // if the outpoint was the last unspent: its height
    int nVersion;         // if the outpoint was the last unspent: its version
```

Reply:
I don't understand your comment here. If `coins->HasUnspent()` is false here, it will not become true in subsequent iterations.

Reply:
Okay, the error was actually on line 2265, that was what confused me.